DeviceCode/Targets/Native/STM32F4/DeviceCode/STM32F4_Analog/STM32F4_AD_functions.cpp

```diff
@@ -46,34 +50,76 @@ static const BYTE g_STM32F4_AD_Pins[] = STM32F4_ADC_PINS;
 
 BOOL AD_Initialize( ANALOG_CHANNEL channel, INT32 precisionInBits )
 {
-    if (!(RCC->APB2ENR & RCC_APB2ENR_ADCxEN)) { // not yet initialized
-        RCC->APB2ENR |= RCC_APB2ENR_ADCxEN; // enable AD clock
-        ADC->CCR = 0; // ADCCLK = PB2CLK / 2;
-        ADCx->SQR1 = 0; // 1 conversion
-        ADCx->CR1 = 0;
-        ADCx->CR2 = ADC_CR2_ADON; // AD on
-        ADCx->SMPR1 = 0x01249249 * STM32F4_AD_SAMPLE_TIME;
-        ADCx->SMPR2 = 0x09249249 * STM32F4_AD_SAMPLE_TIME;
+    int chNum = g_STM32F4_AD_Channel[channel];
+
+    // init this channel if it's listed in the STM32F4_AD_CHANNELS array
+    for (int i = 0; i < STM32F4_AD_NUM ; i++) {
+        if (g_STM32F4_AD_Channel[i] == chNum) {
+            // valid channel
+            if (!(RCC->APB2ENR & RCC_APB2ENR_ADCxEN)) { // not yet initialized
+                RCC->APB2ENR |= RCC_APB2ENR_ADCxEN; // enable AD clock
+                ADC->CCR = 0; // ADCCLK = PB2CLK / 2;
+                ADCx->SQR1 = 0; // 1 conversion
+                ADCx->CR1 = 0;
+                ADCx->CR2 = ADC_CR2_ADON; // AD on
+                ADCx->SMPR1 = 0x01249249 * STM32F4_AD_SAMPLE_TIME;
+                ADCx->SMPR2 = 0x09249249 * STM32F4_AD_SAMPLE_TIME;
+            }
+            
+            // set pin as analog input if channel is not one of the internally connected
+            if(chNum <= 15) {
+                CPU_GPIO_DisablePin(AD_GetPinForChannel(channel), RESISTOR_DISABLED, 0, GPIO_ALT_MODE_1);
+                return TRUE;        
+            }
+        }
     }
-    // set pin as analog input
-    CPU_GPIO_DisablePin(AD_GetPinForChannel(channel), RESISTOR_DISABLED, 0, GPIO_ALT_MODE_1);
-    return TRUE;
+	
+    // channel not available
+    return FALSE;
 }
 
 void AD_Uninitialize( ANALOG_CHANNEL channel )
 {
-    // free pin
-    CPU_GPIO_DisablePin(AD_GetPinForChannel(channel), RESISTOR_DISABLED, 0, GPIO_ALT_PRIMARY);
+    int chNum = g_STM32F4_AD_Channel[channel];
+
+    // free GPIO pin if this channel is listed in the STM32F4_AD_CHANNELS array 
+    // and if it's not one of the internally connected ones as these channels don't take any GPIO pins
+    if(chNum <= 15) {
+        CPU_GPIO_DisablePin(AD_GetPinForChannel(channel), RESISTOR_DISABLED, 0, GPIO_ALT_MODE_1);

```

Comment:
You are right it should revert to GPIO_ALT_PRIMARY.
Just corrected the issue and submitted pull request #89.